### PR TITLE
[SOAR-6617] Fix Get Incident Attachment response type bug

### PR DESCRIPTION
**Pull request by mberezin-r7**
### Description

Describe the proposed changes:

  - This PR fixes a reported issue that the Get Incident Attachment action incorrectly parses a returned attachment type (as raw bytes) as a string.
  - JIRA: https://issues.corp.rapid7.com/browse/SOAR-6617

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing

Successfully decoded file attachment from ServiceNow by passing the attachment contents to the base64 plugin for decoding.

Screen Shot 2021-08-13 at 1 25 46 PM

This proves that actions whose response content is in fact JSON gets parsed correctly.

Screen Shot 2021-08-13 at 3 33 10 PM

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Comment by mberezin-r7** [comment]
This PR is ready for review. The PluginValidator fails only because the plugin.spec.yaml and help.md are missing examples. This is irrelevant to this issue and I recommend that we ignore this validator failure and force merge this PR with reviewer approvals.

**Comment by cmchugh-r7** [comment]
if i'm understanding, seems like the intended behavior is for the attachment output to be bytes, right? if that's the case, should the output type in the plugin spec be bytes instead of string?

File: plugins/servicenow/icon_servicenow/actions/get_incident_attachment/action.py [hunk 1]
```diff
@@ -25,4 +26,4 @@ def run(self, params={}):
 
         response = self.connection.request.make_request(url, method)
 
-        return {Output.ATTACHMENT_CONTENTS: response.get("resource")}
+        return {Output.ATTACHMENT_CONTENTS: str(base64.b64encode(response.get("resource")), "utf-8")}
```
Comment (dgreene-r7) [inline plugins/servicenow/icon_servicenow/actions/get_incident_attachment/action.py:29]: This is going to raise an exception if the status code is a `204` and the `make_request` function returns `resource` as `None`.

**Review by dgreene-r7 — CHANGES_REQUESTED** [review]
1 major issue.

Other than that, I'd look at ensuring that data that's returned from `requests` is bytes-like. To do this, I'd do your base64 encoding in the `make_request` method in a `try/except` block before you return so you can properly handle cases if the response isn't of type `bytes`.

**Comment by mberezin-r7** [comment]
> if i'm understanding, seems like the intended behavior is for the attachment output to be bytes, right? if that's the case, should the output type in the plugin spec be bytes instead of string?

The intended output of this action is the string of the base64 encoding of the attachment bytes. I believe that this is accurately reflected in the current documentation https://github.com/rapid7/insightconnect-plugins/blob/master/plugins/servicenow/help.md#output-5

**Comment by jschipp-r7** [comment]
@mberezin-r7  To echo Charlotte, we call that intended output `bytes` in the product. It looks like it was incorrectly typed previously. `byes` in the product (Not Python bytes type) is a higher-level construct - it is just a literal string data type but one where we expect the format to be in base64 (for validation, etc)

Because the string and bytes output types both end up as string, I think this can be a patch version bump like you have without breaking any workflows.

```
    output:
      attachment_contents:
        title: Attachment Contents
        description: The Base64-encoded contents of the downloaded attachment
        type: bytes
        required: true
```

**Comment by mberezin-r7** [comment]
> @mberezin-r7 To echo @cmchugh-r7 , we call that intended output `bytes` in the product. It looks like it was incorrectly typed previously. `byes` in the product (Not Python bytes type) is a higher-level construct - it is just a literal string data type but one where we expect the format to be in base64 (for validation, etc)
> 
> Because the string and bytes output types both end up as string, _I think_ this can be a patch version bump like you have without breaking any workflows.
> 
> ```
>     output:
>       attachment_contents:
>         title: Attachment Contents
>         description: The Base64-encoded contents of the downloaded attachment
>         type: bytes
>         required: true
> ```

Thanks for the clarification @jschipp-r7, I'll update the Get Incident Attachments action output type to the higher-level `bytes` construct in the plugin spec, test it in the staging workflow screenshotted above, and report back.

**Comment by mberezin-r7** [comment]
After changing the output type to the `bytes` higher-level construct, the Get Incident Attachment continues to work as expected in staging.

Screen Shot 2021-08-16 at 9 37 25 AM

Thanks for the suggestion @cmchugh-r7 and @jschipp-r7 , let me know if there are any other changes recommended prior to merge and release.

**Review by jschipp-r7 — APPROVED** [review]

**Pull request merged**